### PR TITLE
Support watching assets on a specific account

### PR DESCRIPTION
**Pull request by Cal-L**
**Support watching assets on a specific account**

The purpose of the changes to is allow a user to store a watched asset to the token states in the `TokensController`. Why is this needed? The introduction of the Permission System on mobile now allows users to interact with Dapps using accounts that are different than what is active on their wallet screen. This dynamic makes it possible for users to watch assets under the active Dapp account, which may be different than the active wallet account. As a result, we needed to enable passing of an account address to both the `watchAsset` and `addToken` method, that would allow for storing of tokens under specific account addresses. All of the listed changes below is backwards compatible and code that uses the `TokensController` will not need to change.

**Description**

- CHANGED:

  - `TokensController.watchAsset`
    - Added optional `interactingAddress: string` parameter with the definition of `The address of the account that is requesting to watch the asset.`
  - `TokensController.acceptWatchAsset`
    - Now passes an account address (`interactingAddress: string || selectedAddress: string`) into `TokensController.addToken` under the `ERC20` switch condition.
  - `TokensController.addToken`
    -  If `interactingAddress` is defined and is not equal to `selectedAddress`, this method will function the same way but will not update `tokens`, `ignoredTokens`, `detectedTokens` token states.
    - If `interactingAddress` is not defined, this method will use the `selectedAddress` and function the same way.
  - `SuggestedAssetMeta` type
    -  Added optional `interactingAddress?: string` property. This is only populated if `interactingAddress` is passed into `TokensController.watchAsset`.

**Checklist**

- [x] Tests are included if applicable
- [x] Any added code is fully documented

**Issue**

Resolves #???

### CODE REVIEW

File: packages/assets-controllers/src/TokensController.ts [hunk 1]
```diff
@@ -598,16 +621,24 @@ export class TokensController extends BaseController<
...
+    interactingAddress?: string,
+  ): Promise<AssetSuggestionResult> {
+    const suggestedAssetMeta: SuggestedAssetMeta = {
       asset,
       id: this._generateRandomId(),
       status: SuggestedAssetStatus.pending as SuggestedAssetStatus.pending,
       time: Date.now(),
       type,
     };
+    if (interactingAddress) {
```
Comment (Cal-L) [inline packages/assets-controllers/src/TokensController.ts:639]: TODO - Provide selectedAddress as fallback to ensure that suggestedAssetMeta always has an interactingAddress

**Review by tommasini — APPROVED** [review]
LGTM!

File: packages/assets-controllers/src/TokensController.ts (hunk 1 above)
Comment (Gudahtt) [reply packages/assets-controllers/src/TokensController.ts:639]: Hmm, what would the purpose of an interacting address be? I was thinking of the token list as being per-chain, not per chain/address. That seems to be how it's tracked in state at the moment as well. So there would be no way to watch an asset just for a single account.

Comment (Gudahtt) [reply packages/assets-controllers/src/TokensController.ts:639]: Oh I see, that's what this PR is for. Sorry I lost track of the context here for a moment.

**Comment by Gudahtt** [comment]
Hmm, I hadn't realized before that EIP-747 relied upon the current selected account. I guess we'll need to update https://eips.ethereum.org/EIPS/eip-747 to work more effectively with scenarios where multiple accounts are connected.

**Comment by adonesky1** [comment]
> Hmm, I hadn't realized before that EIP-747 relied upon the current selected account. I guess we'll need to update https://eips.ethereum.org/EIPS/eip-747 to work more effectively with scenarios where multiple accounts are connected.

we're discussing related topics in context of adding support for adding NFTs via `wallet_watchAsset` and architecting changes with multichain in mind. @Cal-L perhaps you should join us to discuss this work in our meeting tomorrow?

**Review by Gudahtt — APPROVED** [review]
LGTM! Left one suggestion for improved test coverage, and Cal your pending TODO comment looks worth pursing as well, but we don't need to block on that.

**Comment by Cal-L** [comment]
@metamaskbot publish-preview

File: packages/assets-controllers/src/TokensController.ts [hunk 2]
```diff
@@ -323,16 +334,26 @@ export class TokensController extends BaseController<
...
-        ignoredTokens: newIgnoredTokens,
-        detectedTokens: newDetectedTokens,
+      let newState: Partial<TokensState> = {
         allTokens: newAllTokens,
         allIgnoredTokens: newAllIgnoredTokens,
         allDetectedTokens: newAllDetectedTokens,
-      });
+      };
+
+      // Only update active tokens if user is interacting with their active wallet account.
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply packages/assets-controllers/src/TokensController.ts:346]: Yeah exactly! Thanks

**Pull request merged**